### DISCUSSION TITLE
Is there a way to invalidate session cookie, when refresh token expires

### DISCUSSION THREAD

**Discussion by skatanski** [DuendeSoftware/community#8] [BFF] [answered]
Version: 7.0.8

.net 8

From documentation, I can see it's possible to synchronize refresh token's lifetime with session cookie in a way that clears refresh tokens when session cookie expires. I can't find, if the other way around is possible - to make it so that session cookie expires, when the refresh token expires. We'd like the user on the front-end side to have to manually re-authenticate, when refresh tokens have expired, with this change being triggered by the back-end.

**Comment by skatanski** [comment]
I've investigated it a bit more, and I can see I could implement something like this in the RefreshTokenStore (whenever a refresh token is removed) or in my own implementation of DefaultRefreshTokenService. I could query session store and remove any sessions referenced to a given user/client. But perhaps there's a better way to do it, and if there are any counterpoint/or would it be considered bad practice by adding such functionality in the classes I've mentioned.

**Comment by RolandGuijt** [comment]
Setting the [CoordinateClientLifetimesWithUserSession option](https://docs.duendesoftware.com/identityserver/v7/reference/options/) in combination with server side sessions should work. Both the session lifetime and refresh token lifetime should be the same.

The session by default has sliding expiration. Whenever the refresh token is used, then the session is renewed. So if the refresh token isn't used and expires after X hours, then the session will expire at the same time.
The only case where the session would survive past the refresh token is if the user interacts with the IdentityServer but not the client app.

**Comment by RolandGuijt** [comment]
@skatanski Did my comment make things clear for you? If so I'd like to close the issue.

**Comment by skatanski** [comment]
Hi @RolandGuijt so I've set both:
- identityServerOptoins.Authentication.CoordinateClientLifetimesWithUserSession to true
- client specific field CoordinateLifetimeWithUserSession to true

However, after i run following scenario:
- request access token (lifetime 5 minutes) and single use refresh token (absolute lifetime 10 minutes)
- new access token/refresh token pair gets requested before 5 minute expiry
- next time I request new pair - that fails because rightly refresh token got expired
- I redirect user to login prompt - user gets automatically authenticated with a session cookie

I was expecting that actual login form would show up because the session should have expired along with the refresh token.

**Comment by skatanski** [comment]
Interestingly I've found following issue, which seems to overlap with mine:
https://github.com/DuendeSoftware/Support/issues/391
However `DefaultSessionCoordinationService.ValidateSessionAsync` doesn't seem to run, when an authentication using session cookie happens.

**Comment by skatanski** [comment]
So I've digged a bit more into it, and my current solution to terminate user session on failed refresh token validation is following:
- I've derived DefaultRefreshTokenService with my own implementation
- ValidateRefreshTokenAsync does the following extra, this is a POC:
```
if (result?.Error == "invalid_grant")
        {
            var refreshToken = await RefreshTokenStore.GetRefreshTokenAsync(tokenHandle);
            if(refreshToken != null)
                await serverSideSessionStore.DeleteSessionsAsync(new SessionFilter { SessionId = refreshToken.SessionId, SubjectId = refreshToken.SubjectId });
        }
```
This way whenever absolute refresh token expires, session will get terminated as well. I'd need to add client flag to enable it per client.
@RolandGuijt can you please advise if this is the recommended way, or whether I'm missing something/breaking any good practice

**Comment by RolandGuijt** [comment]
Just to make sure we didn't misunderstand: do you have server-side sessions enabled?

**Comment by RolandGuijt** [comment]
@skatanski Would you like to follow up on this? Or can we close the issue?

**Comment by skatanski** [comment]
Yes I do have server side sessions enabled. Sorry for the delay.

**Comment by maartenba** [comment]
(note: we're moving this issue to our new community discussions)

**Comment by RolandGuijt** [comment, marked as answer]
It looks I might have misunderstood earlier:

If you enable `CoordinateClientLifetimesWithUserSession` it's the lifetime of the session that is leading. When the session ends, the refresh token will expire. And whenever the refresh token is used it will slide the session and extend the lifetime of the session by default.
The feature isn't designed to be used the other way around. Ending the session when the refresh token expires would mean that users are logged out of the SSO session having consequences for all active client sessions.

**Comment by greg-signi** [comment]
@RolandGuijt 
But having the Session active when I can't get new access tokens to target my destination apis, isn't it useless to keep the session alive ?

I've ran into a Zombie session problem, related to the issue the OP was mentioning

**Reply by khalidabuhakmeh** [reply]
Hello Greg. It really depends on the application you're building and the fallback behavior you may want. You might be correct in your particular case, but each application may have different behaviors and outcomes.

Could you describe your particular application?

**Reply by greg-signi** [reply]
Hello, my issue is when the identity server removes the session from his side of the spectre, I would want the same to happen for the BFF. Because although I can't get new access tokens, my session still is alive when calling the `/user` endpoint.

And I would assume that the Session would be removed with this option turned on

`options.RemoveSessionAfterRefreshTokenExpiration = true;`

But as of now this is what I see:

// `Duende.Bff.Yarp.Internal.AccessTokenRequestTransform.cs`
> Failed to request new User Access Token due to: \"not_authenticated\". This likely means that the user's refresh token is expired or revoked. The user's session will be ended, which will force the user to log

// `Duende.Bff.SessionManagement.Configuration.PostConfigureApplicationCookieRevokeRefreshToken.cs`
> Revoking user's refresh tokens in OnSigningOut for subject id: null

Then it goes to the `Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler.cs` and finally says

> AuthenticationScheme: "cookie_dashboard" signed out.

I'm trying to find a way now to remove the session from the BFF cleanly

**Reply by wcabus** [reply]
@greg-signi Have you configured both the BFF and IdentityServer to store server-side sessions and to coordinate the BFF's session with IdentityServer?
1. On IdentityServer, ensure that [server-side sessions](https://docs.duendesoftware.com/identityserver/ui/server-side-sessions/) are enabled and persisted to a store.
1. And at the BFF side, you also need [server-side sessions](https://docs.duendesoftware.com/bff/fundamentals/session/server-side-sessions/) enabled and backed by a store like a SQL database, in order to use the back-channel logout endpoint.
2. In IdentityServer, you need to configure the [BackChannelLogoutUri](https://docs.duendesoftware.com/identityserver/reference/models/client/#authentication--session-management) to point to the `/bff/backchannel` endpoint [of the BFF host](https://docs.duendesoftware.com/bff/fundamentals/session/management/back-channel-logout/).
3. Also in IdentityServer, you either need to enable [CoordinateClientLifetimesWithUserSession](https://docs.duendesoftware.com/identityserver/reference/options/#authentication) at a global level, or [CoordinateLifetimeWithUserSession](https://docs.duendesoftware.com/identityserver/reference/models/client/#authentication--session-management) on the BFF host's client level if you want to opt-in on a per-client basis.
